File: 3. Iterations/iterations2.js
// 2. Done
// Bonus. Further work needed (n=7 in 3s, n=8 in 37s, n=9 in 769s guessing n=10 several hours). I can get down to 200secs for n=9 by calcing 10^dp terms before checking.

"use strict"; /* jshint
node : true
*/

var n = 7,
    pi = function (dp) {
        var sum = 0.0,
            lastSum = 0.0,
            n = 0,
            leibnizTerm = function (n) {
                return Math.pow(-1, n) / (n + n + 1);
            },
            truncate = function (x, precision) {
                return parseInt(x * Math.pow(10, precision)) / Math.pow(10, precision); // Does not round
            };

        while (true) {
            lastSum = truncate(sum, dp);
            sum += 4 * leibnizTerm(n++);
            if (truncate(sum, dp) === lastSum) {return lastSum; }
            // Return the current term if it didn't change to dp decimal places (irrespective of pi(dp-1))
        }
    };

console.log("PI (to " + n + " dp's) = " + pi(n));

File: 3. Iterations/iterations2.go
// 2. Done
// Bonus. Pretty slow so far, 2-3x slower than js :-(

package main

import (
	"fmt"
	"math"
)

func leibnizTerm(n int) float64 {
	return math.Pow(-1, float64(n)) / float64(n+n+1)
}

func truncate(x float64, dp int) float64 {
	return math.Floor(math.Pow(10, float64(dp))*x) / math.Pow(10, float64(dp))
}

func pi(dp int) float64 {
	for n, lastSum, sum := 0, 0.0, 0.0; ; n++ {
		lastSum = truncate(sum, dp) // Truncating now saves me 2 more later.
		sum += 4 * leibnizTerm(n)
		if truncate(sum, dp) == lastSum {
			return lastSum
		}
	}
}

func main() {
	n := 7
	fmt.Printf("PI correct to %d decimal places = %g\n", n, pi(n))
}
